README.md
# edsac-peripherals
Individual parts have their own readmes where you can find bills of materials and suggested suppliers, however all parts need wire, which can be purchased  at a large selection of [online](http://uk.farnell.com/c/cable-wire-cable-assemblies/hook-up-wire-single-conductor?wire-gauge=22awg), [nationanl chain](https://www.maplin.co.uk/p/maplin-06-mm-solid-copper-core-22-awg-2-amp-bell-wire-10m-black-bl85g) and local retailers\.
Our parts lists specify parts required for using ATMega328 \(as used in the Arduino Uno\) microcontrollers to drive the peripherals\, but you can use a wide range of different microcontrollers if you wish\. We used [Embecosm USB to UART converters](www.embecosm.com/resources/hardware/#EHW5), but since these are not widely available\, you will have to use a commercially available alternative if you use a controller without one built in\, of which there are many\.


ChangeLog
2018-02-04  Jeremy Bennett  <jeremy.bennett@embecosm.com>

	This is the definitive paper tape reader as demonstrated at FOSDEM
	2018.  Content is now recognized as EDSAC modified Baudot code
	used for input (note, not the same as the representation for
	output).

	* order-generator/printcode/printcode.ino (void setup): Correctly
	place the leader message.
	* papertape-reader/code/reader/reader.ino (SPROCKET_PIN): Move
	sprocket pin to be A2.
	(seekingOrder, struct ChMap, hash, addHash, lookupOrder)
	(lookupNum, initHash): Created.
	(setup): Call initHash, intialize seekingOrder state
	(loop): Interpret input as EDSAC order codes, printing each
	completed code.

2018-02-02  Jeremy Bennett  <jeremy.bennett@embecosm.com>

	This is the definitive printer.  Actual programs are defined as
	a string of EDSAC order mnemonics towards the top of the program.

	* order-generator/printcode/printcode.ino: Generate EDSAC baudot
	code, recognizing just the EDSAC input character set.

2018-02-02  Jeremy Bennett  <jeremy.bennett@embecosm.com>

	* order-generator/printcode/printcode.ino: Produce output in the
	correct format - move sprocket hole one across.
	* papertape-reader/signal-trace.ods: Final version.

2018-02-02  Jeremy Bennett  <jeremy.bennett@embecosm.com>

	This is the fully working reader of paper tape.  A full data log
	is included for tutuorial purposes.

	At present this reads as plain 5 bit binary.  Future work will add
	correct EDSAC modified baudot decoding.

	A ChangeLog is added now there is significant updating to software
	in progress.

	* ChangeLog: Created.
	* papertape-reader/code/reader/reader.ino: Rewritten for a simpler
	analysis and using the correct line as sprocket.
	* papertape-reader/signal-trace.ods: Created.


order-generator/genmap.c
// Map generator for thermal printer

// Copyright (C) 2017 Embecosm Limited <www.embecosm.com>

// Contributor: Peter Bennett <peter.bennett@embecosm.com>
// Contributor: Dan Gorringe <dan.gorringe@embecosm.com>
// Contributor: Jeremy Bennett <jeremy.bennett@embecosm.com>

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.


// Build this program on your host machine, to create the declaration for an
// image map of a black circle within a square.

// These have to be constant static declarations, so they end up in Flash, not
// RAM in the Arduino.

// Usage

//   genmap <varname> <square-size> <hole-diam> [ <fill-ratio> ]

// varname

//   name of the map (C variable name)

// square-size

//   size of the square (pixels)

// hole-diam

//   diameter of the hole.

// fill-ratio

//   gray ratio, 1 in <fill-ratio> black pixels will be white. An omitted
//   value, or value less than 1  means no black pixels are white.  Note that
//   a value of 1 will mean no pixels are printed at all!

// Error checking is minimal.

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


int
main (int   argc,
      char *argv[])
{
  char *varname;
  int  width;
  int  diam;
  int  fill_ratio = 0;
  int  r;
  int  r_sq;

  int  i, x, y;

  uint8_t *pixmap;
  int map_bytes = (width * width + 7) / 8;

  // Get arguments

  if ((argc < 4) || (argc > 5))
    {
      printf ("Usage: genmap <varname> <square-size> <hole-diam> [ <fill-ratio> ]\n");
      exit (EXIT_FAILURE);
    }

  varname = argv[1];
  width = atoi (argv[2]);
  diam = atoi (argv[3]);

  if (5 == argc)
    fill_ratio = atoi (argv[4]);

  if ((4 == argc) || (fill_ratio < 1))
    fill_ratio = width * width;

  // Compute derived values

  r = diam / 2;
  r_sq = r * r;
  map_bytes = (width * width + 7) / 8;

  // Set up and clear the pixmap

  pixmap = malloc (width * width * sizeof (*pixmap));
  memset (pixmap, 0, map_bytes);

  for (x = 0; x < width; x++)
    {
      int  x_off    = x - width / 2;
      int  x_off_sq = x_off * x_off;

      for (y = 0; y < width; y++)
	{
	  int  y_off = y - width / 2;
	  int  y_off_sq = y_off * y_off;

	  int rowbit  = y * width + x;
	  int rowbyte = rowbit / 8;
	  int bit     = 7 - rowbit % 8;

	  if (((x_off_sq + y_off_sq) < r_sq)
	      && (0 != ((x + y) % fill_ratio)))
	    {
	      // In circle and filled - bit is black.

	      pixmap[rowbyte] |= 1 << bit;
	    }
	  else
	    {
	      // Not in circle or not filled - bit is white

	      pixmap[rowbyte] &= ~(1 << bit);
	    }
	}
    }

  printf ("static const uint8_t  IMAGE_BYTES = %d;\n\n", map_bytes);
  printf ("static const uint8_t  %s[IMAGE_BYTES] = {", varname);

  for (i = 0; i < map_bytes; i++)
    {
      if (0 == i % 8)
	printf ("\n  ");
      else
	printf (" ");

      printf ("0x%02x", pixmap[i]);

      if (i < (map_bytes - 1))
	printf (",");
    }

  printf ("\n};\n");
  free (pixmap);
}


teleprinter-output/uart/chip.v
/******************************************************************************
 *                                                                             *
 * Copyright 2016 myStorm Copyright and related                                *
 * rights are licensed under the Solderpad Hardware License, Version 0.51      *
 * (the “License”); you may not use this file except in compliance with        *
 * the License. You may obtain a copy of the License at                        *
 * http://solderpad.org/licenses/SHL-0.51. Unless required by applicable       *
 * law or agreed to in writing, software, hardware and materials               *
 * distributed under this License is distributed on an “AS IS” BASIS,          *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or             *
 * implied. See the License for the specific language governing                *
 * permissions and limitations under the License.                              *
 *                                                                             *
 ******************************************************************************/

module chip (
	     // 100MHz clock input
	     input 	   clk,
	     // SRAM Memory lines
	     output [18:0] ADR,
	     output [15:0] DAT,
	     output 	   RAMOE,
	     output 	   RAMWE,
	     output 	   RAMCS,
	     // All PMOD outputs
	     output [55:0] PMOD,
	     input [1:0]   BUT,
	     input [3:0]   DIP
	     );

   // SRAM signals are not use in this design, lets set them to default values
   assign ADR[18:0] = {19{1'b0}};
   assign DAT[15:0] = {16{1'b0}};
   assign RAMOE = 1'b1;
   assign RAMWE = 1'b1;
   assign RAMCS = 1'b1;

   assign PMOD[49:0] = {50{1'b0}};

   wire 		   OUT;

   wire 		   enter_ed;
   wire 		   next_ed;

   //UART(/slow) clock

   reg [9:0] 		   clock_divider_counter;
   reg 			   uart_clock;

   always @(posedge clk) begin
      // 	if (reset == 1'b1)
      // 	  clock_divider_counter <= 0;
      if (clock_divider_counter == 434)
   	clock_divider_counter <= 0;
      else
   	// Otherwise increment the counter
   	clock_divider_counter <= clock_divider_counter + 1;
   end

   // Generate a clock (toggle this register)
   always @(posedge clk) begin
      //if (reset == 1'b1)
      //  uart_clock <= 0;
      if (clock_divider_counter == 434)
 	uart_clock <= ~uart_clock;
   end

   uart my_uart (
		 .clk   (uart_clock),
		 .led (PMOD[55:52]),
		 .next_ed (next_ed),
		 .button (BUT[1]),
		 .UART_TX (PMOD[50]),
		 .UART_GND (PMOD[51]) //this is rx?
		 );

   edge_detect next_edge_detect (
				 .clk   (uart_clock),
				 .IN (BUT[0]),
				 .OUT (next_ed)
				 );

endmodule


teleprinter-output/uart/uart.v
//(MINE without buffer)
/******************************************************************************
 *                                                                             *
 * Copyright 2016 myStorm Copyright and related                                *
 * rights are licensed under the Solderpad Hardware License, Version 0.51      *
 * (the “License”); you may not use this file except in compliance with        *
 * the License. You may obtain a copy of the License at                        *
 * http://solderpad.org/licenses/SHL-0.51. Unless required by applicable       *
 * law or agreed to in writing, software, hardware and materials               *
 * distributed under this License is distributed on an “AS IS” BASIS,          *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or             *
 * implied. See the License for the specific language governing                *
 * permissions and limitations under the License.                              *
 *                                                                             *
 ******************************************************************************/

// Copied over uart from original chiphack
// have had little luck in finding the right ports to use on the blackice board
// as cannot seem to use the one programming.

// Start: 8'hD: 8'b00001101
// End: 8'hA: 8'b00001010
// Test: 8'h57: 8'b01010111

module uart(  input        clk,
              input 	   next_ed,
              input 	   button,
              output [3:0] led,        // PMOD[55:52]
              output reg   UART_TX,    // PMOD[50]
              output 	   UART_GND);  // possibly use i_tx(/rx) from PMOD3/4?

   // Some register and wire declarations
   reg [3:0] 		   transmit_state;
   reg [7:0] 		   transmit_data;
   reg 			   key1_reg;
   reg [4:0] 		   word_state;
   reg [3:0] 		   led_value;
   
   //reg [8:0] 		   buffer;
   integer 		   i;
   
   integer 		   len; //length of input string
   reg [120:0] 		   transmit_string; //input string


   // Reset
   wire 		   reset;
   assign reset = ~button;

   // Tie UART_GND low
   assign UART_GND = 0;

   assign led = led_value;

   always @(posedge clk or posedge reset) begin
      if (reset) begin
 	 // Reset to the "IDLE" state
 	 transmit_state <= 0;
	 word_state <= 1;
	 led_value <= 1;
	 
 	 // The UART line is set to '1' when idle, or reset
 	 UART_TX <= 1;

 	 // Data we'll transmit - start at ASCII '0'
 	 transmit_data <= 8'h30;
      end
      else begin
 	 case (transmit_state)
 	   0:
 	     begin
 		// Idle state
		transmit_string = "Hello, World!";
		len = 13;
		led_value <= 1;
			       		
		if (next_ed == 1) begin
		   transmit_state <= 1;
		end
 	     end
 	   1:
 	     begin
 		// Start bit state, and progress onto the next state
		UART_TX <= 0;
		transmit_state <= 2;
		led_value <= 2;
	     end
 	   2,3,4,5,6,7,8,9:
 	     begin
		led_value <= 4;
		
 		// Data bits
		UART_TX <= transmit_data[transmit_state - 2];
		transmit_state <= transmit_state + 1;
 	     end
 	   10:
 	     begin
 		// Fills in transmit_data with next letter
		led_value <= 15;
		UART_TX <= 1;
		transmit_state <= 0; //wait for next button
		if (word_state == 0) begin
		   transmit_state <= 0;
		   word_state <= len + 2;
 		end
		if (word_state == 1) begin
		   word_state <= len + 2;
		   transmit_state <= 0;
		   transmit_data <= 8'hD; //carrage return
		end
		else begin
		   transmit_state <= 1;
		   
		   if (word_state == 1) begin
		      transmit_data <= 8'hA; //newline
		   end
		   else begin
		      i <= word_state - 3;
		      
		      transmit_data[0] <= transmit_string[i*8];
		      transmit_data[1] <= transmit_string[i*8+1];
		      transmit_data[2] <= transmit_string[i*8+2];
		      transmit_data[3] <= transmit_string[i*8+3];
		      transmit_data[4] <= transmit_string[i*8+4];
		      transmit_data[5] <= transmit_string[i*8+5];
		      transmit_data[6] <= transmit_string[i*8+6];
		      transmit_data[7] <= transmit_string[i*8+7];
		      
		      //transmit_state <= 0;
		   end
		   word_state <= word_state - 1;
		end
             end

 	   default:
 	     transmit_state <= 0;
 	 endcase
      end
   end

endmodule
